README.md
[![Build status](https://ci.appveyor.com/api/projects/status/m7s53wav1l0abssg/branch/master?svg=true)](https://ci.appveyor.com/project/bblanchon/arduinojson/branch/master) [![Build Status](https://travis-ci.org/bblanchon/ArduinoJson.svg?branch=master)](https://travis-ci.org/bblanchon/ArduinoJson) [![Coverage Status](https://img.shields.io/coveralls/bblanchon/ArduinoJson.svg)](https://coveralls.io/r/bblanchon/ArduinoJson?branch=master) [![Star this project](http://githubbadges.com/star.svg?user=bblanchon&repo=ArduinoJson&style=flat&color=fff&background=007ec6)](https://github.com/bblanchon/ArduinoJson)

![ArduinoJson's logo](banner.svg)

ArduinoJson - C++ JSON library for IoT
====================

*An elegant and efficient JSON library for embedded systems.*

It's designed to have the most intuitive API, the smallest footprint and is able to work without any allocation on the heap (no malloc).

It has been written with Arduino in mind, but it isn't linked to Arduino libraries so you can use this library in any other C++ project.
For instance, it supports Aduino's `String` and `Stream`, but also `std::string`, `std::istream` and `std::ostream`.

Features
--------

* JSON decoding (comments are supported)
* JSON encoding (with optional indentation)
* Elegant API, very easy to use
* Fixed memory allocation (zero malloc)
* No data duplication (zero copy)
* Portable (written in C++98)
* Self-contained (no external dependency)
* Small footprint
* Header-only library
* MIT License

Works on
--------

* Arduino boards: Uno, Due, Mini, Micro, Yun...
* ESP8266, ESP32
* Teensy
* RedBearLab boards (BLE Nano...)
* Intel Edison and Galileo
* WeMos boards: D1...
* Computers: Windows, Linux, OSX...
* PlatformIO
* Particle
* Energia

Quick start
-----------

#### Decoding / Parsing

```c++
char json[] = "{\"sensor\":\"gps\",\"time\":1351824120,\"data\":[48.756080,2.302038]}";

StaticJsonObject<200> root;

parseJson(root, json);

const char* sensor = root["sensor"];
long time          = root["time"];
double latitude    = root["data"][0];
double longitude   = root["data"][1];
```

[See JsonParserExample.ino](examples/JsonParserExample/JsonParserExample.ino)

Use [ArduinoJson Assistant](https://bblanchon.github.io/ArduinoJson/assistant/) to compute the buffer size.

#### Encoding / Generating

```c++
StaticJsonObject<200> root;
root["sensor"] = "gps";
root["time"] = 1351824120;

JsonArray& data = root.createNestedArray("data");
data.add(48.756080);
data.add(2.302038);

root.printTo(Serial);
// This prints:
// {"sensor":"gps","time":1351824120,"data":[48.756080,2.302038]}
```

[See JsonGeneratorExample.ino](examples/JsonGeneratorExample/JsonGeneratorExample.ino)

Use [ArduinoJson Assistant](https://bblanchon.github.io/ArduinoJson/assistant/) to compute the buffer size.


Documentation
-------------

The documentation is available online in the [ArduinoJson Website](https://bblanchon.github.io/ArduinoJson/).

The [ArduinoJson Assistant](https://bblanchon.github.io/ArduinoJson/assistant/) helps you get started with the library.


Donators
--------

Special thanks to the following persons and companies who made generous donations to the library author:

* Robert Murphy <img alt='USA' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1fa-1f1f8.svg' width='18' height='18'>
* Surge Communications <img alt='USA' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1fa-1f1f8.svg' width='18' height='18'>
* Alex Scott <img alt='United Kingdom' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1ec-1f1e7.svg' width='18' height='18'>
* Firepick Services LLC <img alt='USA' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1fa-1f1f8.svg' width='18' height='18'>
* A B Doodkorte <img alt='Netherlands' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1f3-1f1f1.svg' width='18' height='18'>
* Scott Smith <img alt='USA' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1fa-1f1f8.svg' width='18' height='18'>
* Johann Stieger <img alt='Austria' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1e6-1f1f9.svg' width='18' height='18'>
* Gustavo Donizeti Gini <img alt='Brazil' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1e7-1f1f7.svg' width='18' height='18'>
* Charles-Henri Hallard <img alt='France' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1eb-1f1f7.svg' width='18' height='18'>
* Martijn van den Burg <img alt='Netherlands' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1f3-1f1f1.svg' width='18' height='18'>
* Nick Koumaris <img alt='Greece' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1ec-1f1f7.svg' width='18' height='18'>
* Jon Williams <img alt='USA' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1fa-1f1f8.svg' width='18' height='18'>
* Kestutis Liaugminas <img alt='Lithuania' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1f1-1f1f9.svg' width='18' height='18'>
* Darlington Adibe <img alt='Nigeria' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1f3-1f1ec.svg' width='18' height='18'>
* Yoeri Kroon <img alt='Netherlands' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1f3-1f1f1.svg' width='18' height='18'>
* Andrew Melvin <img alt='United Kingdom' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1ec-1f1e7.svg' width='18' height='18'>
* Doanh Luong <img alt ='Vietnam' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1fb-1f1f3.svg' width='18' height='18'>
* Christoph Schmidt <img alt='Germany' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1e9-1f1ea.svg' width='18' height='18'>
* OpenEVSE LLC <img alt='USA' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1fa-1f1f8.svg' width='18' height='18'>
* Prokhoryatov Alexey <img alt='Russia' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1f7-1f1fa.svg' width='18' height='18'>
* Google Inc. <img alt='USA' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1fa-1f1f8.svg' width='18' height='18'>
* Charles Haynes <img alt='Australia' src='https://d1j8pt39hxlh3d.cloudfront.net/development/emojione/2.2/989/2546.svg' width='18' height='18'>
* Charles Walker <img alt='USA' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1fa-1f1f8.svg' width='18' height='18'>
* Günther Jehle <img alt='Liechtenstein' src='https://cdnjs.cloudflare.com/ajax/libs/emojione/2.1.4/assets/svg/1f1f1-1f1ee.svg' width='18' height='18'>
* Patrick Elliott


---

Found this library useful? Please star this project or [help me back with a donation!](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=donate%40benoitblanchon%2efr&lc=GB&item_name=Benoit%20Blanchon&item_number=Arduino%20JSON&currency_code=EUR&bn=PP%2dDonationsBF%3abtn_donate_LG%2egif%3aNonHosted) :smile:


CHANGELOG.md
ArduinoJson: change log
=======================

v5.11.2
-------

* Added `DynamicJsonArray`, `DynamicJsonObject`, `StaticJsonArray` etc.
* Added `parseJson()`
* Removed `JsonBuffer::parserArray()` and `parseObject()`
* Removed all deprecated features.
* Fixed `DynamicJsonBuffer::clear()` not resetting allocation size (issue #561)
* Fixed incorrect rounding for float values (issue #588)

### BREAKING CHANGES :warning:

#### Creating a JsonObject

```c++
// Old syntax:
StaticJsonBuffer<200> jb;
JsonObject& obj = jb.createObject();

// New syntax:
StaticJsonObject<200> obj;
```

#### Parsing a JsonObject

```c++
// Old syntax:
StaticJsonBuffer<200> jb;
JsonObject& obj = jb.parseObject(input);

// New syntax:
StaticJsonObject<200> obj;
bool success = parseJson(obj, input);
```

v5.11.1
-------

* Removed dependency on `PGM_P` as Particle 0.6.2 doesn't define it (issue #546)
* Fixed warning "dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]"
* Fixed warning "floating constant exceeds range of 'float' [-Woverflow]" (issue #544)
* Fixed warning "this statement may fall through" [-Wimplicit-fallthrough=] (issue #539)
* Removed `ARDUINOJSON_DOUBLE_IS_64BITS` as it became useless.
* Fixed too many decimals places in float serialization (issue #543)

v5.11.0
-------

* Made `JsonBuffer` non-copyable (PR #524 by @luisrayas3)
* Added `StaticJsonBuffer::clear()`
* Added `DynamicJsonBuffer::clear()`

v5.10.1
-------

* Fixed IntelliSense errors in Visual Micro (issue #483)
* Fixed compilation in IAR Embedded Workbench (issue #515)
* Fixed reading "true" as a float (issue #516)
* Added `ARDUINOJSON_DOUBLE_IS_64BITS`
* Added `ARDUINOJSON_EMBEDDED_MODE`

v5.10.0
-------

* Removed configurable number of decimal places (issues #288, #427 and #506)
* Changed exponentiation thresholds to `1e7` and `1e-5` (issues #288, #427 and #506)
* `JsonVariant::is<double>()` now returns `true` for integers
* Fixed error `IsBaseOf is not a member of ArduinoJson::TypeTraits` (issue #495)
* Fixed error `forming reference to reference` (issue #495)

### BREAKING CHANGES :warning:

| Old syntax                      | New syntax          |
|---------------------------------|---------------------|
| `double_with_n_digits(3.14, 2)` | `3.14`              |
| `float_with_n_digits(3.14, 2)`  | `3.14f`             |
| `obj.set("key", 3.14, 2)`       | `obj["key"] = 3.14` |
| `arr.add(3.14, 2)`              | `arr.add(3.14)`     |

| Input     | Old output | New output |
|-----------|------------|------------|
| `3.14159` | `3.14`     | `3.14159`  |
| `42.0`    | `42.00`    | `42`       |
| `0.0`     | `0.00`     | `0`        |

| Expression                     | Old result | New result |
|--------------------------------|------------|------------|
| `JsonVariant(42).is<int>()`    | `true`     | `true`     |
| `JsonVariant(42).is<float>()`  | `false`    | `true`     |
| `JsonVariant(42).is<double>()` | `false`    | `true`     |


v5.9.0
------

* Added `JsonArray::remove(iterator)` (issue #479)
* Added `JsonObject::remove(iterator)`
* Renamed `JsonArray::removeAt(size_t)` into `remove(size_t)`
* Renamed folder `include/` to `src/`
* Fixed warnings `floating constant exceeds range of float`and `floating constant truncated to zero` (issue #483)
* Removed `Print` class and converted `printTo()` to a template method (issue #276)
* Removed example `IndentedPrintExample.ino`
* Now compatible with Particle 0.6.1, thanks to Jacob Nite (issue #294 and PR #461 by @foodbag)

v5.8.4
------

* Added custom implementation of `strtod()` (issue #453)
* Added custom implementation of `strtol()` (issue #465)
* `char` is now treated as an integral type (issue #337, #370)

v5.8.3
------

* Fixed an access violation in `DynamicJsonBuffer` when memory allocation fails (issue #433)
* Added operators `==` and `!=` for two `JsonVariant`s (issue #436)
* Fixed `JsonVariant::operator[const FlashStringHelper*]` (issue #441)

v5.8.2
------

* Fixed parsing of comments (issue #421)
* Fixed ignored `Stream` timeout (issue #422)
* Made sure we don't read more that necessary (issue #422)
* Fixed error when the key of a `JsonObject` is a `char[]` (issue #423)
* Reduced code size when using `const` references
* Fixed error with string of type `unsigned char*` (issue #428)
* Added `deprecated` attribute on `asArray()`, `asObject()` and `asString()` (issue #420)

v5.8.1
------

* Fixed error when assigning a `volatile int` to a `JsonVariant` (issue #415)
* Fixed errors with Variable Length Arrays (issue #416)
* Fixed error when both `ARDUINOJSON_ENABLE_STD_STREAM` and `ARDUINOJSON_ENABLE_ARDUINO_STREAM` are set to `1`
* Fixed error "Stream does not name a type" (issue #412)

v5.8.0
------

* Added operator `==` to compare `JsonVariant` and strings (issue #402)
* Added support for `Stream` (issue #300)
* Reduced memory consumption by not duplicating spaces and comments

### BREAKING CHANGES :warning:

`JsonBuffer::parseObject()` and  `JsonBuffer::parseArray()` have been pulled down to the derived classes `DynamicJsonBuffer` and `StaticJsonBufferBase`.

This means that if you have code like:

```c++
void myFunction(JsonBuffer& jsonBuffer);
```

you need to replace it with one of the following:

```c++
void myFunction(DynamicJsonBuffer& jsonBuffer);
void myFunction(StaticJsonBufferBase& jsonBuffer);
template<typename TJsonBuffer> void myFunction(TJsonBuffer& jsonBuffer);
```


v5.7.3
------

* Added an `printTo(char[N])` and `prettyPrintTo(char[N])` (issue #292)
* Added ability to set a nested value like this: `root["A"]["B"] = "C"` (issue #352)
* Renamed `*.ipp` to `*Impl.hpp` because they were ignored by Arduino IDE (issue #396)

v5.7.2
------

* Made PROGMEM available on more platforms (issue #381)
* Fixed PROGMEM causing an exception on ESP8266 (issue #383)

examples.js
var examples = {
  "arduinoJson": {
    "sensor": "gps",
    "time": 1351824120,
    "data": [
      48.756080,
      2.302038
    ]
  },
  "openweathermap": {
    "coord": {
      "lon": -0.13,
      "lat": 51.51
    },
    "weather": [
      {
        "id": 301,
        "main": "Drizzle",
        "description": "drizzle",
        "icon": "09n"
      },
      {
        "id": 701,
        "main": "Mist",
        "description": "mist",
        "icon": "50n"
      },
      {
        "id": 741,
        "main": "Fog",
        "description": "fog",
        "icon": "50n"
      }
    ],
    "base": "stations",
    "main": {
      "temp": 281.87,
      "pressure": 1032,
      "humidity": 100,
      "temp_min": 281.15,
      "temp_max": 283.15
    },
    "visibility": 2900,
    "wind": {
      "speed": 1.5
    },
    "clouds": {
      "all": 90
    },
    "dt": 1483820400,
    "sys": {
      "type": 1,
      "id": 5091,
      "message": 0.0226,
      "country": "GB",
      "sunrise": 1483776245,
      "sunset": 1483805443
    },
    "id": 2643743,
    "name": "London",
    "cod": 200
  },
  "wunderground": {
    "response": {
      "version": "0.1",
      "termsofService": "http://www.wunderground.com/weather/api/d/terms.html",
      "features": {
        "conditions": 1
      }
    },
    "current_observation": {
      "image": {
        "url": "http://icons-ak.wxug.com/graphics/wu2/logo_130x80.png",
        "title": "Weather Underground",
        "link": "http://www.wunderground.com"
      },
      "display_location": {
        "full": "San Francisco, CA",
        "city": "San Francisco",
        "state": "CA",
        "state_name": "California",
        "country": "US",
        "country_iso3166": "US",
        "zip": "94101",
        "latitude": "37.77500916",
        "longitude": "-122.41825867",
        "elevation": "47.00000000"
      },
      "observation_location": {
        "full": "SOMA - Near Van Ness, San Francisco, California",
        "city": "SOMA - Near Van Ness, San Francisco",
        "state": "California",
        "country": "US",
        "country_iso3166": "US",
        "latitude": "37.773285",
        "longitude": "-122.417725",
        "elevation": "49 ft"
      },
      "estimated": {},
      "station_id": "KCASANFR58",
      "observation_time": "Last Updated on June 27, 5:27 PM PDT",
      "observation_time_rfc822": "Wed, 27 Jun 2012 17:27:13 -0700",
      "observation_epoch": "1340843233",
      "local_time_rfc822": "Wed, 27 Jun 2012 17:27:14 -0700",
      "local_epoch": "1340843234",
      "local_tz_short": "PDT",
      "local_tz_long": "America/Los_Angeles",
      "local_tz_offset": "-0700",
      "weather": "Partly Cloudy",
      "temperature_string": "66.3 F (19.1 C)",
      "temp_f": 66.3,
      "temp_c": 19.1,
      "relative_humidity": "65%",
      "wind_string": "From the NNW at 22.0 MPH Gusting to 28.0 MPH",
      "wind_dir": "NNW",
      "wind_degrees": 346,
      "wind_mph": 22.0,
      "wind_gust_mph": "28.0",
      "wind_kph": 35.4,
      "wind_gust_kph": "45.1",
      "pressure_mb": "1013",
      "pressure_in": "29.93",
      "pressure_trend": "+",
      "dewpoint_string": "54 F (12 C)",
      "dewpoint_f": 54,
      "dewpoint_c": 12,
      "heat_index_string": "NA",
      "heat_index_f": "NA",
      "heat_index_c": "NA",
      "windchill_string": "NA",
      "windchill_f": "NA",
      "windchill_c": "NA",
      "feelslike_string": "66.3 F (19.1 C)",
      "feelslike_f": "66.3",
      "feelslike_c": "19.1",
      "visibility_mi": "10.0",
      "visibility_km": "16.1",
      "solarradiation": "",
      "UV": "5",
      "precip_1hr_string": "0.00 in ( 0 mm)",
      "precip_1hr_in": "0.00",
      "precip_1hr_metric": " 0",
      "precip_today_string": "0.00 in (0 mm)",
      "precip_today_in": "0.00",
      "precip_today_metric": "0",
      "icon": "partlycloudy",
      "icon_url": "http://icons-ak.wxug.com/i/c/k/partlycloudy.gif",
      "forecast_url": "http://www.wunderground.com/US/CA/San_Francisco.html",
      "history_url": "http://www.wunderground.com/history/airport/KCASANFR58/2012/6/27/DailyHistory.html",
      "ob_url": "http://www.wunderground.com/cgi-bin/findweather/getForecast?query=37.773285,-122.417725"
    }
  }
}
